contributingGuides/STYLE.md

```diff
@@ -511,112 +958,215 @@ In React Native, one **must not** attempt to falsey-check a string for an inline
-        ...
-        return <input {...props} ref={ref} />
-    }
+```tsx
+import type {ForwarderRef} from 'react';
 
-    export default React.forwardRef(FancyInput)
+type FancyInputProps = {
+    ...
+};
+
+function FancyInput(props: FancyInputProps, ref: ForwardedRef<TextInput>) {
+    ...
+    return <input {...props} ref={ref} />
+};
+
+export default React.forwardRef(FancyInput)
 ```
 
-## Stateless components vs Pure Components vs Class based components vs Render Props - When to use what?
+If the ref handle is not available (e.g. `useImperativeHandle` is used) you can define a custom handle type above the component.
 
-Class components are DEPRECATED. Use function components and React hooks.
+```tsx 
+import type {ForwarderRef} from 'react';
+import {useImperativeHandle} from 'react';
 
-[https://react.dev/reference/react/Component#migrating-a-component-with-lifecycle-methods-from-a-class-to-a-function](https://react.dev/reference/react/Component#migrating-a-component-with-lifecycle-methods-from-a-class-to-a-function)
+type FancyInputProps = {
+    ...
+    onButtonPressed: () => void;
+};
 
-## Composition vs Inheritance
+type FancyInputHandle = {
+  onButtonPressed: () => void;
+}
 
-From React's documentation -
->Props and composition give you all the flexibility you need to customize a component’s look and behavior in an explicit and safe way. Remember that components may accept arbitrary props, including primitive values, React elements, or functions.
->If you want to reuse non-UI functionality between components, we suggest extracting it into a separate JavaScript module. The components may import it and use that function, object, or a class, without extending it.
+function FancyInput(props: FancyInputProps, ref: ForwardedRef<FancyInputHandle>) {
+    useImperativeHandle(ref, () => ({onButtonPressed}));
 
-Use an HOC a.k.a. *[Higher order component](https://reactjs.org/docs/higher-order-components.html)* if you find a use case where you need inheritance.
+    ...
+    return <input {...props} ref={ref} />;
+};
 
-If several HOC need to be combined, there is a `compose()` utility. But we should not use this utility when there is only one HOC.
+export default React.forwardRef(FancyInput)
+```
 
-```javascript
-// Bad
-export default compose(
-    withLocalize,
-)(MyComponent);
+### Hooks and HOCs
+
+Use hooks whenever possible, avoid using HOCs.
+
+> Why? Hooks are easier to use (can be used inside the function component), and don't need nesting or `compose` when exporting the component. It also allows us to remove `compose` completely in some components since it has been bringing up some issues with TypeScript. Read the [`compose` usage](#compose-usage) section for further information about the TypeScript issues with `compose`.
+
+> Note: Because Onyx doesn't provide a hook yet, in a component that accesses Onyx data with `withOnyx` HOC, please make sure that you don't use other HOCs (if applicable) to avoid HOC nesting.
```

Comment:
This is out of date - let's update this to formalize  `useOnyx` as a best practice over `withOnyx`